File: backend-go/internal/model/procurement.go
package model

import "time"

// أنواع طلبات المشتريات — كل نوع له صلاحية مستقلة يمنحها الأدمن يدوياً للموظف
// (procurement_personal / procurement_customer)، ما تنجر تلقائياً مع أي دور.
const (
	RequestTypePersonalSupply  = "PERSONAL_SUPPLY"  // طلب احتياجات شخصية للموظف نفسه
	RequestTypeCustomerProduct = "CUSTOMER_PRODUCT" // طلب منتج للزبون (الفلو الأصلي، مربوط بحجز)
	// طلب مادة يدوي: الإداري أو الليدر يكتب المادة بالتفصيل — موديل،
	// نوعية، عدد، سعر تقريبي — بدون ما يكون مربوط بحجز ولا بكتالوك.
	RequestTypeManualSupply = "MANUAL_SUPPLY"
)

var RequestTypeLabels = map[string]string{
	RequestTypePersonalSupply:  "احتياجات شخصية",
	RequestTypeCustomerProduct: "منتج للزبون",
	RequestTypeManualSupply:    "طلب مادة يدوي",
}

type ProcurementRequest struct {
	ID               string   `db:"id" json:"id"`
	Code             string   `db:"code" json:"code"`
	RequestedByID    string   `db:"requestedById" json:"-"`
	BookingID        *string  `db:"bookingId" json:"-"`
	RequestType      string   `db:"requestType" json:"requestType"`
	Notes            *string  `db:"notes" json:"notes"`
	Status           string   `db:"status" json:"status"`
	FulfilledByID    *string  `db:"fulfilledById" json:"-"`
	TotalCost        *float64 `db:"totalCost" json:"totalCost"`
	FulfillmentNotes *string  `db:"fulfillmentNotes" json:"fulfillmentNotes"`
	// المورد الي انجابت منه المادة — يتحدد وقت التجهيز، إلزامي حتى
	// نعرف من وين انجابت ونكدر نحاسب على السعر.
	SupplierID  *string    `db:"supplierId" json:"supplierId"`
	CreatedAt   time.Time  `db:"createdAt" json:"createdAt"`
	FulfilledAt *time.Time `db:"fulfilledAt" json:"fulfilledAt"`

	Items       []ProcurementItem          `db:"-" json:"items"`
	RequestedBy *ProcurementRequesterBrief `db:"-" json:"requestedBy"`
	FulfilledBy *EmployeeIDNameBrief       `db:"-" json:"fulfilledBy"`
	Booking     *ProcurementBookingBrief   `db:"-" json:"booking"`
	Supplier    *EmployeeIDNameBrief       `db:"-" json:"supplier"`
	TypeLabel   string                     `db:"-" json:"typeLabel"`
}

// ProcurementRequesterBrief يماثل { select: { id: true, name: true, role: true } } بالباك إند القديم
type ProcurementRequesterBrief struct {
	ID   string `db:"id" json:"id"`
	Name string `db:"name" json:"name"`
	Role string `db:"role" json:"role"`
}

// EmployeeIDNameBrief يماثل { select: { id: true, name: true } } بالباك إند القديم
type EmployeeIDNameBrief struct {
	ID   string `db:"id" json:"id"`
	Name string `db:"name" json:"name"`
}

// ProcurementBookingBrief يماثل include: { booking: { include: { customer: { select ... } } } } بالباك إند القديم
type ProcurementBookingBrief struct {
	ID string `db:"id" json:"id"`
	// كود الحجز — إداري الكميات يحتاجه حتى يعرف أي حجز يخص الطلب.
	// كانت الواجهة تعرضه وهو مو موجود بالرد أصلاً، فيطلع «undefined».
	Code     string         `db:"code" json:"code"`
	Customer *CustomerBrief `db:"-" json:"customer"`
}

type CustomerBrief struct {
	ID    string `db:"id" json:"id"`
	Name  string `db:"name" json:"name"`
	Phone string `db:"phone" json:"phone"`
}

type ProcurementItem struct {
	ID          string   `db:"id" json:"id"`
	RequestID   string   `db:"requestId" json:"requestId"`
	ProductName string   `db:"productName" json:"productName"`
	Quantity    int      `db:"quantity" json:"quantity"`
	UnitPrice   *float64 `db:"unitPrice" json:"unitPrice"`
	TotalPrice  *float64 `db:"totalPrice" json:"totalPrice"`
	Fulfilled   bool     `db:"fulfilled" json:"fulfilled"`
	// تفاصيل الطلب اليدوي — الموديل والنوعية وملاحظات المادة
	Model     *string `db:"model" json:"model"`
	Spec      *string `db:"spec" json:"spec"`
	ItemNotes *string `db:"itemNotes" json:"itemNotes"`
}

type CreateProcurementItemRequest struct {
	ProductName string   `json:"productName"`
	Quantity    int      `json:"quantity"`
	UnitPrice   *float64 `json:"unitPrice"`
	TotalPrice  *float64 `json:"totalPrice"`
	Model       *string  `json:"model"`
	Spec        *string  `json:"spec"`
	ItemNotes   *string  `json:"itemNotes"`
}

type CreateProcurementRequestRequest struct {
	RequestedByID string                         `json:"requestedById"`
	BookingID     *string                        `json:"bookingId"`
	RequestType   string                         `json:"requestType"`
	Notes         *string                        `json:"notes"`
	Items         []CreateProcurementItemRequest `json:"items"`
}

type UpdateProcurementStatusRequest struct {
	Status string `json:"status"`
}

type FulfillProcurementItemRequest struct {
	ID         string   `json:"id"`
	UnitPrice  *float64 `json:"unitPrice"`
	TotalPrice *float64 `json:"totalPrice"`
	Fulfilled  *bool    `json:"fulfilled"`
}

type FulfillProcurementRequestRequest struct {
	FulfilledByID    string                          `json:"fulfilledById"`
	SupplierID       *string                         `json:"supplierId"`
	TotalCost        *float64                        `json:"totalCost"`
	FulfillmentNotes *string                         `json:"fulfillmentNotes"`
	Items            []FulfillProcurementItemRequest `json:"items"`
}

type ProcurementStats struct {
	TotalSpent     float64            `json:"totalSpent"`
	TotalItems     int                `json:"totalItems"`
	PendingCount   int                `json:"pendingCount"`
	MonthlySpent   float64            `json:"monthlySpent"`
	FulfilledCount int                `json:"fulfilledCount"`
	ByMonth        map[string]float64 `json:"byMonth"`
}

File: backend/src/routes/procurement.ts
import { Router } from 'express'
import { prisma } from '../prisma'
import { requireRole } from '../middleware/requireAuth'

const router = Router()

// GET / - list all procurement requests
router.get('/', async (_req, res) => {
  const requests = await prisma.procurementRequest.findMany({
    include: {
      items: true,
      requestedBy: { select: { id: true, name: true, role: true } },
      fulfilledBy: { select: { id: true, name: true } },
      booking: { include: { customer: { select: { id: true, name: true, phone: true } } } },
    },
    orderBy: { createdAt: 'desc' },
  })
  res.json(requests)
})

// GET /stats - procurement statistics
router.get('/stats', async (_req, res) => {
  const now = new Date()
  const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1)

  const [totalSpent, totalItems, pendingCount, monthlySpent, fulfilled] = await Promise.all([
    prisma.procurementRequest.aggregate({
      _sum: { totalCost: true },
      where: { status: 'FULFILLED' },
    }),
    prisma.procurementItem.count(),
    prisma.procurementRequest.count({ where: { status: 'PENDING' } }),
    prisma.procurementRequest.aggregate({
      _sum: { totalCost: true },
      where: { status: 'FULFILLED', fulfilledAt: { gte: startOfMonth } },
    }),
    prisma.procurementRequest.count({ where: { status: 'FULFILLED' } }),
  ])

  // Monthly breakdown (last 6 months)
  const sixMonthsAgo = new Date(now.getFullYear(), now.getMonth() - 5, 1)
  const monthlyData = await prisma.procurementRequest.findMany({
    where: { status: 'FULFILLED', fulfilledAt: { gte: sixMonthsAgo } },
    select: { totalCost: true, fulfilledAt: true },
  })

  const byMonth: Record<string, number> = {}
  for (const r of monthlyData) {
    if (!r.fulfilledAt) continue
    const key = `${r.fulfilledAt.getFullYear()}-${String(r.fulfilledAt.getMonth() + 1).padStart(2, '0')}`
    byMonth[key] = (byMonth[key] || 0) + (r.totalCost || 0)
  }

  res.json({
    totalSpent: totalSpent._sum.totalCost || 0,
    totalItems,
    pendingCount,
    monthlySpent: monthlySpent._sum.totalCost || 0,
    fulfilledCount: fulfilled,
    byMonth,
  })
})

// POST / - create a new procurement request
router.post('/', async (req, res) => {
  const { requestedById, bookingId, notes, items } = req.body

  // Auto-generate code PR-XXXX
  const last = await prisma.procurementRequest.findFirst({
    orderBy: { code: 'desc' },
    select: { code: true },
  })
  let nextNum = 1
  if (last?.code) {
    const match = last.code.match(/PR-(\d+)/)
    if (match) nextNum = parseInt(match[1], 10) + 1
  }
  const code = `PR-${String(nextNum).padStart(4, '0')}`

  const request = await prisma.procurementRequest.create({
    data: {
      code,
      requestedById,
      bookingId: bookingId || null,
      notes: notes || null,
      items: {
        create: (items || []).map((item: any) => ({
          productName: item.productName,
          quantity: item.quantity,
          unitPrice: item.unitPrice || null,
          totalPrice: item.totalPrice || null,
        })),
      },
    },
    include: {
      items: true,
      requestedBy: { select: { id: true, name: true, role: true } },
      booking: { include: { customer: { select: { id: true, name: true, phone: true } } } },
    },
  })
  res.status(201).json(request)
})

// PUT /:id/status - update status
router.put('/:id/status', requireRole('ADMIN', 'FINANCE'), async (req, res) => {
  const { id } = req.params
  const { status } = req.body

  const request = await prisma.procurementRequest.update({
    where: { id },
    data: { status },
    include: {
      items: true,
      requestedBy: { select: { id: true, name: true, role: true } },
      fulfilledBy: { select: { id: true, name: true } },
      booking: { include: { customer: { select: { id: true, name: true, phone: true } } } },
    },
  })
  res.json(request)
})

// PUT /:id/fulfill - mark as fulfilled with costs
router.put('/:id/fulfill', requireRole('ADMIN', 'FINANCE'), async (req, res) => {
  const { id } = req.params
  const { fulfilledById, totalCost, fulfillmentNotes, items } = req.body

  // Update individual item costs if provided
  if (items && Array.isArray(items)) {
    for (const item of items) {
      await prisma.procurementItem.update({
        where: { id: item.id },
        data: {
          unitPrice: item.unitPrice ?? undefined,
          totalPrice: item.totalPrice ?? undefined,
          fulfilled: item.fulfilled ?? true,
        },
      })
    }
  }

  const request = await prisma.procurementRequest.update({
    where: { id },
    data: {
      status: 'FULFILLED',
      fulfilledById,
      totalCost: totalCost || null,
      fulfillmentNotes: fulfillmentNotes || null,
      fulfilledAt: new Date(),
    },
    include: {
      items: true,
      requestedBy: { select: { id: true, name: true, role: true } },
      fulfilledBy: { select: { id: true, name: true } },
      booking: { include: { customer: { select: { id: true, name: true, phone: true } } } },
    },
  })
  res.json(request)
})

export default router
